Core/clim-core/gadgets.lisp

```diff
@@ -1164,9 +1164,9 @@ and must never be nil.")
 
 (defmethod handle-event ((pane push-button-pane) (event pointer-button-release-event))
   (with-slots (armed pressedp) pane
-    (setf pressedp nil)
-    (when armed
-      (activate-callback pane (gadget-client pane) (gadget-id pane))
+    (when (and armed pressedp)
```

Comment:
Fixing the issue mentioned in the commit message is good, of course. 

This commit has another side-effect: since `pressedp` now stays true until after `activate-callback` returns, so the button is rendered as pressed while `activate-callback` is executing. Pressing the button again can cause the pressed state to be reset when one of the `activate-callback`s returns, leading to an inconsistency. For this new behavior to work, we would have to count active "presses" for the button, I think.

Reply:
So how about this?

```diff
modified   Core/clim-core/gadgets.lisp
@@ -1128,7 +1128,7 @@ and must never be nil.")
                              activate/deactivate-repaint-mixin
                              enter/exit-arms/disarms-mixin
                              sheet-leaf-mixin)
-  ((pressedp          :initform nil)
+  ((pressed-count     :initform 0)
    (show-as-default-p :type boolean
                       :initform nil
                       :initarg :show-as-default-p
@@ -1158,26 +1158,26 @@ and must never be nil.")
                          :height 2*border-thickness)))
 
 (defmethod handle-event ((pane push-button-pane) (event pointer-button-press-event))
-  (with-slots (pressedp) pane
-    (setf pressedp t)
+  (with-slots (pressed-count) pane
+    (incf pressed-count)
     (dispatch-repaint pane +everywhere+)))
 
 (defmethod handle-event ((pane push-button-pane) (event pointer-button-release-event))
-  (with-slots (armed pressedp) pane
-    (when (and armed pressedp)
-      (activate-callback pane (gadget-client pane) (gadget-id pane)))
-    (when pressedp
-      (setf pressedp nil)
-      (dispatch-repaint pane +everywhere+))))
+  (with-slots (armed pressed-count) pane
+    (when (plusp pressed-count)
+      (when armed
+        (activate-callback pane (gadget-client pane) (gadget-id pane)))
+      (when (zerop (decf pressed-count))
+        (dispatch-repaint pane +everywhere+)))))
 
 (defmethod handle-repaint ((pane push-button-pane) region)
   (declare (ignore region))
-  (with-slots (armed pressedp) pane
+  (with-slots (pressed-count) pane
     (with-bounding-rectangle* (x1 y1 x2 y2) (sheet-region pane)
       (draw-rectangle* pane x1 y1 x2 y2 :ink (effective-gadget-background pane))
       (draw-bordered-rectangle* pane x1 y1 x2 y2
                                 :border-width 1
-                                :style (if (and pressedp armed) :inset :outset))
+                                :style (if (plusp pressed-count) :inset :outset))
       (let* ((x-spacing (pane-x-spacing pane))
              (y-spacing (pane-y-spacing pane))
              (border-thickness *3d-border-thickness*)
```

Reply:
couldn't we get away with:
```
(when pressedp
  (setf pressedp nil)
  (when armed (activate-callback))
  (dispatch-repaint))
```
?

Reply:
Yes, rearranging the body seems best. The "pressed count" idea would have been nice, but it actually doesn't work with the way command loops nest.

Reply:
fix pushed